Add support for some smartthings and xiaomi devices

damarco opened the pull request.

rcloran commented:
Thanks! This is awesome. I'll work through this today -- there are some changes (eg, signature to list) which I can kinda see the point of, but aren't used, so I'm not sure whether we should go for the simplest possible thing now or not.

damarco commented:
The idea behind the list is, that there might me multiple generations of a device with slightly differing signatures, but all could be handled by the same custom device, eg. [smartthings moisture sensor](https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/smartsense-moisture-sensor.src/smartsense-moisture-sensor.groovy) .

I have also added a last_seen attribute to a device, which is not really related to the quirks branch, but I have used this to derive an alive/missing status.

In your initial commit the custom device logic was not applied when the device was loaded from the db, so i guess there are two ways to handle this. Either persist the device class somehow in the db, or apply the quirks.get_device() logic after loading the original device. Not sure which one is better.

I have added a custom_device_registry to handle manufacturer specific clusters in the zcl.deserialize() method, which would also require a change to bellows to pass the device object.

Also the Hue and Tradfri handlers are no longer needed, because the zll/zha profile id issue has already been fixed upstream.

rcloran commented:
I've updated the quirks branch to use something a lot like your database logic (I changed device_save to a listener signal called raw_device_initialized). The appdb code is pretty janky, and I don't think trying to do anything fancier would be great. Thanks for the inspiration :)

I'm not going to merge this PR to the quirks branch. Here's the parts of it that I've identified and what I've done with them:

 - last_seen: in zigpy master
 - Database fixes: As above
 - BatteryCustomDevice: I'm going to leave that out in the first merge of quirks. That's not to say we shouldn't have it, I just want to keep it as simple as I can get away with.
 - New custom devices: Will put into quirks (without battery bits)
 - Custom cluster registry: I want to think on how this should be done. I think would like to just get rid of the cluster registry idea and delegate deserialization down the tree, like it is done for messages.

If there's anything I'm missing, please let me know.

Thanks again for the extra custom devices and the fixes!
